### PR TITLE
Fixes a bug when running queries with a limit clause

### PR DESCRIPTION
**Pull request by adarshsanjeev**
When creating the queryDefinition and calculating shuffleSpecs, we defaulted to a shuffling everything into a single partition if we required a limit(regardless of the resultShuffleSpecFactory passed to the queryKit). This was followed by the limit stage having no shuffling (as it would always create a single output partition).

This logic is correct, as long as we do not require any specific partitioning for the final output, since the limit would be the final MSQ stage(this would be the case for instance, in the case of writing SELECT query results into a query report).

However, for query types which expect partitions of a specific size, this causes a bug. For example, an async query which applies a LIMIT of 1000, with a context parameter `rowsPerPage:100`. This would expect the limit processor to create a 10 segments of size 100 each. However, the limit processor would still create a single segment as an output. This causes issues with certain queries which would expect multiple partitions.

This issue is present with all `SELECT` queries except ones which write to the query report. `REPLACE` or `INSERT` queries which have an ALL partitioning are also affected (since we do not allow LIMIT along with other partitions).

Even worse, we still sometimes expect these partitions, and this causes an exception when the subsequent stage tries to read them.

This PR aims to correct this by adding a shuffling based on the resultShuffleSpecFactory after a limit processor depending on the query destination. LimitFrameProcessors currently do not update the partition boosting column, so we also add the boost column to the previous stage, if one is required.

<h2>Behaviour changes</h2>

**INSERT + LIMIT with PARTITIONED BY ALL**

Before: Only the scan stage would have a shuffle spec (mix). Would create a single segment. `rowsPerSegment` would not be respected.

After: Scan stage has mix shuffle, limit stage has sort shuffle spec. `rowsPerSegment` parameter of 1000 is respected. If we have a clustering, the first scan stage is a sort shuffle spec instead.
Screenshot 2024-07-05 at 9 45 40 PM

**INSERT + LIMIT with PARTITIONED BY DAY (if we supported it)**

Before: Only the scan stage would have a shuffle spec (mix, this would ignore day partitioning). Would create a single segment. `rowsPerSegment` would not be respected.
After: Scan stage has mix shuffle, limit stage has sort shuffle spec. `rowsPerSegment` parameter of 1000 is respected. (probably)
**SELECT * FROM foo LIMIT 100 -> taskReport**
Before: Works as expected.
After: Works as expected.
Screenshot 2024-07-05 at 9 49 56 PM

**SELECT * FROM foo LIMIT 100 -> durableStorage (async queries)**
Before: Only the scan stage would have a shuffle spec (mix). Would create a single segment. `rowsPerPage` would not be respected.
After: The scan stage has a mix shuffle spec. The limit stage has a sort with target size shuffle, determined by `rowsPerPage`. We have an additional selectResults stage
Screenshot 2024-07-05 at 9 51 48 PM

**SELECT * FROM foo LIMIT 100 ORDER BY dim1 -> durableStorage (async queries)**
Before: Only the scan stage would have a shuffle spec (mix). Would create a single segment. `rowsPerPage` would not be respected.
After: The scan stage has a sort shuffle spec with target count of worker count. The limit stage has a sort with target size shuffle, determined by `rowsPerPage`. We have an additional selectResults stage.
Screenshot 2024-07-05 at 9 54 17 PM
**SELECT * FROM foo LIMIT 100 -> export**
Before: Only the scan stage would have a shuffle spec (mix). Would create a single segment. `rowsPerPage` would not be respected.
After: The scan stage has a mix shuffle spec. The limit stage has a sort with target size shuffle, determined by `rowsPerPage`. 
Screenshot 2024-07-05 at 9 57 41 PM
**SELECT * FROM (SELECT * FROM foo ORDER BY dim1 LIMIT 10000) ORDER BY dim2 LIMIT 5000 with 1 worker**
Before: Works as expected.
After: Works as expected.
Screenshot 2024-07-05 at 10 00 24 PM

**SELECT * FROM (SELECT * FROM foo ORDER BY dim1 LIMIT 10000) ORDER BY dim2 LIMIT 5000 with 2 worker**
Before: Inner limit produces single partition on one of the workers, other workers produce nothing on the limit. The next stage throws an exception when another worker tries to read one of these missing partitions.
```
2024-07-05T14:02:38,587 INFO [controller-status-checker-0] org.apache.druid.msq.exec.WorkerImpl - Controller task[query-3eacaa7e-d5a0-434e-934a-d319751fd6a4] for worker task[query-3eacaa7e-d5a0-434e-934a-d319751fd6a4-worker1_0] failed. Canceling.
2024-07-05T14:02:38,589 INFO [task-runner-0-priority-0] org.apache.druid.indexing.worker.executor.ExecutorLifecycle - Task completed with status: {
  "id" : "query-3eacaa7e-d5a0-434e-934a-d319751fd6a4-worker1_0",
  "status" : "FAILED",
  "duration" : 1319,
  "errorMsg" : "WorkerRpcFailed: RPC call to task failed unrecoverably: [query-3eacaa7e-d5a0-434e-934a-d319751fd6a4-worker0_0]",
  "location" : {
    "host" : null,
    "port" : -1,
    "tlsPort" : -1
  }
}
```
After: Works as expected. The scan stages have shuffles with `target number of partitions=1` (to handle the sorting), the inner limit stage has a sort with `target number of partitions=3(for worker count)`, and the final limit has sort with `targetSize = 1000 (rowsPerPage)`. We have a selectResults stage since we are doing a sort in the previous stage.
Screenshot 2024-07-05 at 10 01 53 PM

**INSERT + Select dim1, count(*) GROUP BY 1 LIMIT 5000 PARTITIONED BY ALL with rowsPerSegment:1000**
Before: The group by stages have globalSort and null shuffle specs. The limit stage creates a single partition.
Screenshot 2024-07-06 at 9 19 31 PM
After: The limit stage has sort shuffle spec into targetSize.`rowsPerSegment` parameter of 1000 is respected.
Screenshot 2024-07-06 at 9 44 50 PM

**GroupBy -> taskReport**
Before: Works as expected.
After: Works as expected.
Screenshot 2024-07-06 at 9 58 37 PM

**GroupBy -> durableStorage (asyncQuery)**
Before: The group by stages have globalSort and null shuffle specs. The limit stage creates a single partition. Would create a single page. `rowsPerPage` would not be respected.
After: The limit stage has a sort shuffle spec into targetSize. We always add an additional stage `selectResults`. `rowsPerPage` parameter of 50 is respected.
Screenshot 2024-07-06 at 9 59 08 PM

**GroupBy -> export**
Before: The group by stages have globalSort and null shuffle specs. The limit stage creates a single partition. Would create a single page. `rowsPerPage` would not be respected.
After: The limit stage has a sort shuffle spec into targetSize. `rowsPerPage` parameter of 50 is respected.
Screenshot 2024-07-06 at 10 01 34 PM

<hr>

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Review by LakshSingla — COMMENTED** [review]
Thanks for the patch. The finalShuffleSpec must also be set in the case of ingest queries. Otherwise, MSQ will disregard the `rowsPerSegment` parameter, if we have a LIMIT clause.

File: extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/test/MSQTestBase.java [hunk 1]
```diff
@@ -1285,12 +1285,10 @@ public void verifyResults()
...
                                                                       .collect(Collectors.toList());
-            if (diskSegmentList.size() != 1) {
-              throw new IllegalStateException("Single key in multiple partitions");
-            }
-            SegmentId diskSegment = diskSegmentList.get(0);
             // Checking if the row belongs to the correct segment interval
-            Assert.assertTrue(segmentIdVsOutputRowsMap.get(diskSegment).contains(Arrays.asList(row)));
+            Assert.assertTrue(diskSegmentList.stream()
+                                             .map(segmentIdVsOutputRowsMap::get)
+                                             .anyMatch(rows -> rows.contains(Arrays.asList(row))));
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/test/MSQTestBase.java:1291]: 1. Each row is present in exactly one interval
2. Each row is present in exactly one segment. 


This is the behavior we want.

Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/test/MSQTestBase.java]: Added an additional filter here to check that it only matches a certain segment and interval.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/groupby/GroupByQueryKit.java [hunk 2]
```diff
@@ -132,12 +133,12 @@ public QueryDefinition makeQueryDefinition(
...
                                          ? ShuffleSpecFactories.singlePartition()
                                          : ShuffleSpecFactories.globalSortWithMaxPartitionCount(maxWorkerCount);
-      shuffleSpecFactoryPostAggregation = doLimitOrOffset
+      shuffleSpecFactoryPostAggregation = doLimitOrOffset && !needsFinalShuffling
                                           ? ShuffleSpecFactories.singlePartition()
                                           : resultShuffleSpecFactory;
       partitionBoost = true;
     } else {
-      shuffleSpecFactoryPreAggregation = doLimitOrOffset
+      shuffleSpecFactoryPreAggregation = doLimitOrOffset && !needsFinalShuffling
```
Comment (LakshSingla) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/groupby/GroupByQueryKit.java:141]: This change looks incorrect - Irrespective of whether we require final shuffling or not, to compute the limit properly, we need to generate a single partition.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/groupby/GroupByQueryKit.java [hunk 3]
```diff
@@ -132,12 +133,12 @@ public QueryDefinition makeQueryDefinition(
       shuffleSpecFactoryPreAggregation = intermediateClusterBy.isEmpty()
                                          ? ShuffleSpecFactories.singlePartition()
                                          : ShuffleSpecFactories.globalSortWithMaxPartitionCount(maxWorkerCount);
-      shuffleSpecFactoryPostAggregation = doLimitOrOffset
+      shuffleSpecFactoryPostAggregation = doLimitOrOffset && !needsFinalShuffling
```
Comment (LakshSingla) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/groupby/GroupByQueryKit.java:136]: Same reason as on the other line - Consider a query like: 

```sql
SELECT col1 FROM foo GROUP BY col1 LIMIT 1
```

If the group by returns 10 partitions, the limit would work on them individually, which is not desirable. I think this change isn't required as we are always setting up limit at the end.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/groupby/GroupByQueryKit.java [hunk 4]
```diff
@@ -185,13 +186,16 @@ public QueryDefinition makeQueryDefinition(
       );
 
       if (doLimitOrOffset) {
+        final ShuffleSpec finalShuffleSpec = needsFinalShuffling ?
+                                             shuffleSpecFactoryPreAggregation.build(resultClusterBy, true) :
+                                             null;
```
Comment (LakshSingla) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/groupby/GroupByQueryKit.java:191]: I wonder if we can relax this constraint even when we don't have a final shuffle. In intermediate queries, having more partitions will improve the parallelization. Why do we not partition further in case of LIMIT processor?

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java [hunk 5]
```diff
@@ -119,7 +120,8 @@ public QueryDefinition makeQueryDefinition(
     // We ignore the resultShuffleSpecFactory in case:
     //  1. There is no cluster by
     //  2. There is an offset which means everything gets funneled into a single partition hence we use MaxCountShuffleSpec
-    if (queryToRun.getOrderBys().isEmpty() && hasLimitOrOffset) {
+    //  3. The destination does not require shuffling after the limit stage to get outputs of a specific size.
+    if (queryToRun.getOrderBys().isEmpty() && hasLimitOrOffset && !needsFinalShuffling) {
```
Comment (LakshSingla) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java:124]: Why is this change required?

Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java:124]: If this is not checked, it would short circuit to using the MixShuffleSpec. We need to add the boost column, create the clusterBy etc for the scan stage, if we intend to have a final shuffle, so this check should be required.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/groupby/GroupByQueryKit.java (hunk 3 above)
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/groupby/GroupByQueryKit.java:136]: Since it is a subquery, the needsFinalShuffling should always be false here, so I believe that case should still work.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/QueryKit.java [hunk 6]
```diff
@@ -40,13 +40,15 @@ public interface QueryKit<QueryType extends Query<?>>
    * @param minStageNumber           lowest stage number to use for any generated stages. Useful if the resulting
    *                                 {@link QueryDefinition} is going to be added to an existing
    *                                 {@link org.apache.druid.msq.kernel.QueryDefinitionBuilder}.
+   * @param needsFinalShuffling      Determines if a final shuffling is required.
```
[the remark this answers is not in the document]
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/QueryKit.java]: Added a comment

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQControllerTask.java
[the remark this answers is not in the document]
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQControllerTask.java:1]: Added

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/groupby/GroupByQueryKit.java (hunk 3 above)
Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/groupby/GroupByQueryKit.java:136]: My example was incorrect, but I am still doubtful. The only reason that this change seems incorrect is what happens if `doLimitOrOffset=true` but the factory doesn't generate a single partition. 
for example: `EXPORT ... SELECT col1 FROM foo GROUP BY col1 LIMIT 1`
In that case, the LIMIT wouldn't be correct

Limit processor must always receive a single partition, and having `ShuffleSpecFactories.singlePartition` ensures that. The earlier code ensured that. The new code doesn't ensure that, and it can at best have the same result as the old code, or be incorrect (as is my guess). Alternatively, in which case will this change be useful over the original code.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/groupby/GroupByQueryKit.java [hunk 7]
```diff
@@ -185,13 +186,16 @@ public QueryDefinition makeQueryDefinition(
       );
 
       if (doLimitOrOffset) {
+        final ShuffleSpec finalShuffleSpec = needsFinalShuffling ?
+                                             shuffleSpecFactoryPreAggregation.build(resultClusterBy, true) :
```
Comment (LakshSingla) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/groupby/GroupByQueryKit.java:190]:
```suggestion
                                             resultFactory.build(resultClusterBy, true) :
```

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/groupby/GroupByQueryKit.java (hunk 2 above)
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/groupby/GroupByQueryKit.java]: Removed this parameter

**Comment by adarshsanjeev** [comment]
I have refactored the code in the QueryKits and added additional tests to address the comments.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java [hunk 8]
```diff
@@ -185,7 +178,7 @@ public QueryDefinition makeQueryDefinition(
                          .inputs(new StageInputSpec(firstStageNumber))
                          .signature(signatureToUse)
                          .maxWorkerCount(1)
-                         .shuffleSpec(null) // no shuffling should be required after a limit processor.
+                         .shuffleSpec(finalShuffleSpec) // Apply the final shuffling after limit spec.
```
Comment (LakshSingla) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java:181]: This seems like a lot of shuffling. Is there any way we can avoid reshuffling the data by the same cluster by, and just repartition? Perhaps not without any supersorter changes, but I wanted to confirm.

Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java:181]: Well maybe we can preserve the optimisation that we had earlier - if there's no `orderBy` and if `doLimitOrOffset == true`, we don't need to partition boost the intermediate shuffleSpec. 

Actually, we don't need to partitionBoost the intermediate shuffle spec in any case (i.e. the shuffleSpec for the scan stage if there's a limit present) - Since it's all going into a single partition anyway, the partitionBoost won't have any use. 
If there's a limit present, only the final stage should have partition boosting.

File: extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/MSQSelectTest.java [hunk 9]
```diff
@@ -1699,6 +1719,166 @@ public void testScanWithMultiValueSelectQuery(String contextName, Map<String, Ob
...
+            2, 0, "input0"
+        )
+        .setExpectedCountersForStageWorkerChannel(
+            CounterSnapshotMatcher
+                .with().rows(4),
+            2, 0, "output"
+        )
+        .setExpectedCountersForStageWorkerChannel(
+            CounterSnapshotMatcher
+                .with().rows(!context.containsKey(MultiStageQueryContext.CTX_ROWS_PER_PAGE) ? new long[] {4} : new long[] {2, 2}),
```
Comment (LakshSingla) [inline extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/MSQSelectTest.java:1874]: This assertion looks indeterministic. Why is that so?

File: extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/MSQSelectTest.java [hunk 10]
```diff
@@ -1699,6 +1719,166 @@ public void testScanWithMultiValueSelectQuery(String contextName, Map<String, Ob
...
+        .setExpectedRowSignature(expectedResultSignature)
+        .setQueryContext(context)
+        .setExpectedResultRows(ImmutableList.of(
+            new Object[]{"1", 1L}
+        )).verifyResults();
+  }
+
+  @MethodSource("data")
+  @ParameterizedTest(name = "{index}:with context {0}")
+  public void testGroupByWithLimitAndOrdering(String contextName, Map<String, Object> context)
```
Comment (LakshSingla) [inline extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/MSQSelectTest.java:1770]: Shouldn't we have modified the context to generate a single row per partition?

**Review by LakshSingla — COMMENTED** [review]
Final comments after the last change. The patch looks good now.

File: extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/MSQSelectTest.java (hunk 9 above)
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/MSQSelectTest.java:1874]: If the rowsPerPage is set, the counters would be split accordingly, it is deterministic, some of the parameters for the test have this value set, and some don't

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java (hunk 8 above)
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java:181]: The initial ScanFP is the one who increments the boost column, if we do not apply the boost column at that stage, the limit processor output would have boost columns of 0, which can't be split.
I guess additional changes would be needed to allow boosting to work with LimitProcessors before this optimization can be made.

File: extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/MSQSelectTest.java (hunk 10 above)
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/MSQSelectTest.java:1770]: There are context parameters with the rowsPerPages already set. Both branches should be tested here.

File: extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/MSQSelectTest.java (hunk 9 above)
Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/MSQSelectTest.java:1874]: Thanks for the explanation

**Review by LakshSingla — APPROVED** [review]
LGTM, thanks for the patch!

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQControllerTask.java [hunk 11]
```diff
@@ -305,16 +306,38 @@ public Optional<Resource> getDestinationResource()
...
+   */
   public static boolean isExport(final MSQSpec querySpec)
   {
     return querySpec.getDestination() instanceof ExportMSQDestination;
   }
 
+  /**
+   * Checks whether the task is an async query which writes frame files containing the final results into durable storage.
+   */
+  public static boolean writeResultsToDurableStorage(final MSQSpec querySpec)
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQControllerTask.java:328]:
```suggestion
  public static boolean writeFinalStageResultsToDurableStorage(final MSQSpec querySpec)
```

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQControllerTask.java [hunk 12]
```diff
@@ -305,16 +306,38 @@ public Optional<Resource> getDestinationResource()
...
+  {
+    return querySpec.getDestination() instanceof DurableStorageMSQDestination;
+  }
+
+  /**
+   * Checks whether the task is an async query which writes frame files containing the final results into durable storage.
+   */
+  public static boolean writeResultsToTaskReport(final MSQSpec querySpec)
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQControllerTask.java:336]:
```suggestion
  public static boolean writeFinalResultsToTaskReport(final MSQSpec querySpec)
```

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java (hunk 8 above)
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java:181]: Could you please explain the reasoning about this change with the below query shapes. What are the new new stages added . 

* q like ` insert + limit  with partitioned by all`
* q like `insert + limit with partitioned by day` assuming validations on planning side is removed. 
* q like `select * from foo group by a limit 100`-> taskReport
* q like `select * from foo limit 100` -> durableStorage
* q like `select * from foo limit 100 order by x`
* q like `export select * from foo limit 100`
* a inner select query.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/groupby/GroupByQueryKit.java [hunk 13]
```diff
@@ -185,13 +185,14 @@ public QueryDefinition makeQueryDefinition(
       );
 
       if (doLimitOrOffset) {
+        final ShuffleSpec finalShuffleSpec = resultShuffleSpecFactory.build(resultClusterBy, false);
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/groupby/GroupByQueryKit.java:188]: Assume select q's are group by q's for this comment : 
https://github.com/apache/druid/pull/16643#discussion_r1665416378

**Review by cryptoe — COMMENTED** [review]
@adarshsanjeev Please update the description with before and after changes to the number of stages and what kind of local testing is done.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java (hunk 8 above)
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java:197]: Updated the PR description

**Review by asdf2014 — APPROVED** [review]
:rocket:

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java [hunk 14]
```diff
@@ -159,14 +159,30 @@ public QueryDefinition makeQueryDefinition(
...
-    // If there is no limit spec, apply the final shuffling here itself. This will ensure partition sizes etc are respected.
-    // If there is a limit spec, it would write everything into one partition anyway, so the final shuffling should be
-    // applied after that.
+    ShuffleSpec scanShuffleSpec;
+    if (!hasLimitOrOffset) {
+      // If there is no limit spec, apply the final shuffling here itself. This will ensure partition sizes etc are respected.
+      scanShuffleSpec = finalShuffleSpec;
+    } else {
+      // If there is a limit spec, check if there are any non-boost columns to sort in.
+      boolean requiresSort = clusterByColumns.stream()
```
Comment (LakshSingla) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java:168]: Do we need to see if we are sorting on the boost column? Isn't that automatically added by the ScanQueryKit et al?

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java [hunk 15]
```diff
@@ -159,14 +159,30 @@ public QueryDefinition makeQueryDefinition(
...
+                                             .anyMatch(keyColumn -> !QueryKitUtils.PARTITION_BOOST_COLUMN.equals(keyColumn.columnName()));
+      if (requiresSort) {
+        // If yes, do a sort into a single partition.
+        scanShuffleSpec = ShuffleSpecFactories.singlePartition().build(clusterBy, false);
+      } else {
+        // If the only clusterBy column is the boost column, we just use a mix shuffle to avoid unused shuffling.
+        // Note that we still need the boost column to be present in the row signature, since the limit stage would
+        // need it to be populated to do its own shuffling later.
```
Comment (LakshSingla) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java:176]: I was under the impression that the limit factory couldn't partition boost. Am I mistaken?

Comment (LakshSingla) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java:176]: otoh, if there isn't any partition boosting, then even the original code would have run into similar problem if there wasn't a cluster key right?

Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java:176]: The limit factory does not increment the partition boosting at regular intervals. The value would be all 0 if it was not added to the row signature at the scan stage.

> even the original code would have run into similar problem 

The original code would have mix shuffle speced. It would always make a single partition, so yes, it should have been too large.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java (hunk 14 above)
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java:168]: We have a if-branch for window functions above this bit of code, which adds its own cluster bys, which are not dependent on the orderBy.

**Review by LakshSingla — APPROVED** [review]
